In [81]:
from dotenv import load_dotenv
from spotipy.oauth2 import SpotifyClientCredentials
import spotipy
import pandas as pd 
import sqlite3
from sqlite3 import Error

# Load environment variables 
load_dotenv()
sp = spotipy.Spotify(client_credentials_manager=SpotifyClientCredentials())

In [82]:
fav_artists = [
    "Drake",
    "Becky G",
    "Jack Harlow",
    "Karol G",
    "Bad Bunny",
    "J Balvin",
    "Grupo Firme",
    "Queen",
    "Lynyrd Skynyrd",
    "Nirvana",
    "Led Zeppelin",
    "Billie Eilish",
    "J. Cole",
    "Rihanna",
    "Foo Fighters",
    "Natti Natasha",
    "Black Sabbath",
    "Van Halen",
    "Pink Floyd",
    "Britney Spears"
]

In [83]:
def check_if_valid_data(df: pd.DataFrame) -> bool:
    if df.empty:
        print("No data available")
        return False 
    
    if df.isnull().values.any():
        num_nulls = df.isnull().sum().sum()
        raise Exception(f'{num_nulls} null values found')
    
    return True

<h1>ARTISTS</h1>

In [84]:
artist_response = []

for artist in fav_artists:
        search_artist = sp.search(artist, limit=1, type="artist")

        artist_response.append(search_artist)

artist_df = pd.DataFrame.from_dict(artist_response)
artist_df

artists
0   {'href': 'https://api.spotify.com/v1/search?qu...
1   {'href': 'https://api.spotify.com/v1/search?qu...
2   {'href': 'https://api.spotify.com/v1/search?qu...
3   {'href': 'https://api.spotify.com/v1/search?qu...
4   {'href': 'https://api.spotify.com/v1/search?qu...
5   {'href': 'https://api.spotify.com/v1/search?qu...
6   {'href': 'https://api.spotify.com/v1/search?qu...
7   {'href': 'https://api.spotify.com/v1/search?qu...
8   {'href': 'https://api.spotify.com/v1/search?qu...
9   {'href': 'https://api.spotify.com/v1/search?qu...
10  {'href': 'https://api.spotify.com/v1/search?qu...
11  {'href': 'https://api.spotify.com/v1/search?qu...
12  {'href': 'https://api.spotify.com/v1/search?qu...
13  {'href': 'https://api.spotify.com/v1/search?qu...
14  {'href': 'https://api.spotify.com/v1/search?qu...
15  {'href': 'https://api.spotify.com/v1/search?qu...
16  {'href': 'https://api.spotify.com/v1/search?qu...
17  {'href': 'https://api.spotify.com/v1/search?qu...
18  {'href': 'https://api.spotify.com/v1/search?qu...
19  {'href': 'https://api.spotify.com/v1/search?qu...

In [85]:
artists_list = artist_df['artists'].tolist()
artists_list

[{'href': 'https://api.spotify.com/v1/search?query=Drake&type=artist&offset=0&limit=1',
  'items': [{'external_urls': {'spotify': 'https://open.spotify.com/artist/3TVXtAsR1Inumwj472S9r4'},
    'followers': {'href': None, 'total': 67311074},
    'genres': ['canadian hip hop',
     'canadian pop',
     'hip hop',
     'rap',
     'toronto rap'],
    'href': 'https://api.spotify.com/v1/artists/3TVXtAsR1Inumwj472S9r4',
    'id': '3TVXtAsR1Inumwj472S9r4',
    'images': [{'height': 640,
      'url': 'https://i.scdn.co/image/ab6761610000e5eb4293385d324db8558179afd9',
      'width': 640},
     {'height': 320,
      'url': 'https://i.scdn.co/image/ab676161000051744293385d324db8558179afd9',
      'width': 320},
     {'height': 160,
      'url': 'https://i.scdn.co/image/ab6761610000f1784293385d324db8558179afd9',
      'width': 160}],
    'name': 'Drake',
    'popularity': 95,
    'type': 'artist',
    'uri': 'spotify:artist:3TVXtAsR1Inumwj472S9r4'}],
  'limit': 1,
  'next': 'https://api.spotify.c

<h3>That was a mess to go through so I thought it would be easier to get each artist's info one at a time and add it to the dataframe since there were only 20 artists and the layout of the info was consistent.</h3>

In [86]:
artist_response = []

for artist in fav_artists:
        search_artist = sp.search(artist, limit=1, type="artist")
        access_artist_info = search_artist["artists"]["items"][0]

        # Save the data to dictionary
        artist_info = {
            "artist_id": access_artist_info["id"],
            "artist_name": access_artist_info["name"],
            "external_url": access_artist_info["external_urls"]["spotify"],
            "genre": access_artist_info["genres"][0],
            "image_url": access_artist_info["images"][0]["url"],
            "followers": access_artist_info["followers"]["total"],
            "popularity": access_artist_info["popularity"],
            "type": access_artist_info["type"],
            "artist_uri": access_artist_info["uri"]
        }

        artist_response.append(artist_info)

artist_df = pd.DataFrame.from_dict(artist_response)
artist_df


artist_id     artist_name  \
0   3TVXtAsR1Inumwj472S9r4           Drake   
1   4obzFoKoKRHIphyHzJ35G3         Becky G   
2   2LIk90788K0zvyj2JJVwkJ     Jack Harlow   
3   790FomKkXshlbRYZFtlgla         KAROL G   
4   4q3ewBCX7sLwd24euuV69X       Bad Bunny   
5   1vyhD5VmyZ7KMfW5gqLgo5        J Balvin   
6   1dKdetem2xEmjgvyymzytS     Grupo Firme   
7   1dfeR4HaWDbWqFHLkxsg1d           Queen   
8   4MVyzYMgTwdP7Z49wAZHx0  Lynyrd Skynyrd   
9   6olE6TJLqED3rqDCT0FyPh         Nirvana   
10  36QJpDe2go2KgaRleHCDTp    Led Zeppelin   
11  6qqNVTkY8uBg9cP3Jd7DAH   Billie Eilish   
12  6l3HvQ5sa6mXTsMTB19rO5         J. Cole   
13  5pKCCKE2ajJHZ9KAiaK11H         Rihanna   
14  7jy3rLJdDQY21OgRLCZ9sD    Foo Fighters   
15  1GDbiv3spRmZ1XdM1jQbT7   Natti Natasha   
16  5M52tdBnJaKSvOpJGz8mfZ   Black Sabbath   
17  2cnMpRsOVqtPMfq7YiFE6K       Van Halen   
18  0k17h0D3J5VfsdmQ1iZtE9      Pink Floyd   
19  26dSoYclwsYLMAKD3tpOr4  Britney Spears   

                                         external_url  \
0   https://open.spotify.com/artist/3TVXtAsR1Inumw...   
1   https://open.spotify.com/artist/4obzFoKoKRHIph...   
2   https://open.spotify.com/artist/2LIk90788K0zvy...   
3   https://open.spotify.com/artist/790FomKkXshlbR...   
4   https://open.spotify.com/artist/4q3ewBCX7sLwd2...   
5   https://open.spotify.com/artist/1vyhD5VmyZ7KMf...   
6   https://open.spotify.com/artist/1dKdetem2xEmjg...   
7   https://open.spotify.com/artist/1dfeR4HaWDbWqF...   
8   https://open.spotify.com/artist/4MVyzYMgTwdP7Z...   
9   https://open.spotify.com/artist/6olE6TJLqED3rq...   
10  https://open.spotify.com/artist/36QJpDe2go2Kga...   
11  https://open.spotify.com/artist/6qqNVTkY8uBg9c...   
12  https://open.spotify.com/artist/6l3HvQ5sa6mXTs...   
13  https://open.spotify.com/artist/5pKCCKE2ajJHZ9...   
14  https://open.spotify.com/artist/7jy3rLJdDQY21O...   
15  https://open.spotify.com/artist/1GDbiv3spRmZ1X...   
16  https://open.spotify.com/artist/5M52tdBnJaKSvO...   
17  https://open.spotify.com/artist/2cnMpRsOVqtPMf...   
18  https://open.spotify.com/artist/0k17h0D3J5Vfsd...   
19  https://open.spotify.com/artist/26dSoYclwsYLMA...   

                       genre  \
0           canadian hip hop   
1                  dance pop   
2   deep underground hip hop   
3                  reggaeton   
4                  reggaeton   
5                  reggaeton   
6            musica mexicana   
7               classic rock   
8                 album rock   
9                     grunge   
10                album rock   
11                   art pop   
12         conscious hip hop   
13             barbadian pop   
14         alternative metal   
15             latin hip hop   
16                album rock   
17                album rock   
18                album rock   
19                 dance pop   

                                            image_url  followers  popularity  \
0   https://i.scdn.co/image/ab6761610000e5eb429338...   67311074          95   
1   https://i.scdn.co/image/ab6761610000e5eb978f96...   12370513          80   
2   https://i.scdn.co/image/ab6761610000e5ebed3a04...    2565812          83   
3   https://i.scdn.co/image/ab6761610000e5ebb8b788...   28811597          87   
4   https://i.scdn.co/image/ab6761610000e5eb8ee9a6...   56583528         100   
5   https://i.scdn.co/image/ab6761610000e5eb5d0fc4...   35379200          88   
6   https://i.scdn.co/image/ab6761610000e5eb41c3d8...    7420265          82   
7   https://i.scdn.co/image/b040846ceba13c3e9c125d...   42889908          84   
8   https://i.scdn.co/image/ab6761610000e5eb451eb0...    4270893          72   
9   https://i.scdn.co/image/84282c28d851a700132356...   16011536          81   
10  https://i.scdn.co/image/207803ce008388d3427a68...   12588906          77   
11  https://i.scdn.co/image/ab6761610000e5ebd8b998...   69271495          89   
12  https://i.scdn.co/image/ab6761610000e5ebadd503...   17967112          85   
13  https://i.scdn.co/image/ab6761610000e5eb019d68...   502

In [87]:
# remove duplicates 
artist_df.drop_duplicates(subset="artist_id")

artist_id     artist_name  \
0   3TVXtAsR1Inumwj472S9r4           Drake   
1   4obzFoKoKRHIphyHzJ35G3         Becky G   
2   2LIk90788K0zvyj2JJVwkJ     Jack Harlow   
3   790FomKkXshlbRYZFtlgla         KAROL G   
4   4q3ewBCX7sLwd24euuV69X       Bad Bunny   
5   1vyhD5VmyZ7KMfW5gqLgo5        J Balvin   
6   1dKdetem2xEmjgvyymzytS     Grupo Firme   
7   1dfeR4HaWDbWqFHLkxsg1d           Queen   
8   4MVyzYMgTwdP7Z49wAZHx0  Lynyrd Skynyrd   
9   6olE6TJLqED3rqDCT0FyPh         Nirvana   
10  36QJpDe2go2KgaRleHCDTp    Led Zeppelin   
11  6qqNVTkY8uBg9cP3Jd7DAH   Billie Eilish   
12  6l3HvQ5sa6mXTsMTB19rO5         J. Cole   
13  5pKCCKE2ajJHZ9KAiaK11H         Rihanna   
14  7jy3rLJdDQY21OgRLCZ9sD    Foo Fighters   
15  1GDbiv3spRmZ1XdM1jQbT7   Natti Natasha   
16  5M52tdBnJaKSvOpJGz8mfZ   Black Sabbath   
17  2cnMpRsOVqtPMfq7YiFE6K       Van Halen   
18  0k17h0D3J5VfsdmQ1iZtE9      Pink Floyd   
19  26dSoYclwsYLMAKD3tpOr4  Britney Spears   

                                         external_url  \
0   https://open.spotify.com/artist/3TVXtAsR1Inumw...   
1   https://open.spotify.com/artist/4obzFoKoKRHIph...   
2   https://open.spotify.com/artist/2LIk90788K0zvy...   
3   https://open.spotify.com/artist/790FomKkXshlbR...   
4   https://open.spotify.com/artist/4q3ewBCX7sLwd2...   
5   https://open.spotify.com/artist/1vyhD5VmyZ7KMf...   
6   https://open.spotify.com/artist/1dKdetem2xEmjg...   
7   https://open.spotify.com/artist/1dfeR4HaWDbWqF...   
8   https://open.spotify.com/artist/4MVyzYMgTwdP7Z...   
9   https://open.spotify.com/artist/6olE6TJLqED3rq...   
10  https://open.spotify.com/artist/36QJpDe2go2Kga...   
11  https://open.spotify.com/artist/6qqNVTkY8uBg9c...   
12  https://open.spotify.com/artist/6l3HvQ5sa6mXTs...   
13  https://open.spotify.com/artist/5pKCCKE2ajJHZ9...   
14  https://open.spotify.com/artist/7jy3rLJdDQY21O...   
15  https://open.spotify.com/artist/1GDbiv3spRmZ1X...   
16  https://open.spotify.com/artist/5M52tdBnJaKSvO...   
17  https://open.spotify.com/artist/2cnMpRsOVqtPMf...   
18  https://open.spotify.com/artist/0k17h0D3J5Vfsd...   
19  https://open.spotify.com/artist/26dSoYclwsYLMA...   

                       genre  \
0           canadian hip hop   
1                  dance pop   
2   deep underground hip hop   
3                  reggaeton   
4                  reggaeton   
5                  reggaeton   
6            musica mexicana   
7               classic rock   
8                 album rock   
9                     grunge   
10                album rock   
11                   art pop   
12         conscious hip hop   
13             barbadian pop   
14         alternative metal   
15             latin hip hop   
16                album rock   
17                album rock   
18                album rock   
19                 dance pop   

                                            image_url  followers  popularity  \
0   https://i.scdn.co/image/ab6761610000e5eb429338...   67311074          95   
1   https://i.scdn.co/image/ab6761610000e5eb978f96...   12370513          80   
2   https://i.scdn.co/image/ab6761610000e5ebed3a04...    2565812          83   
3   https://i.scdn.co/image/ab6761610000e5ebb8b788...   28811597          87   
4   https://i.scdn.co/image/ab6761610000e5eb8ee9a6...   56583528         100   
5   https://i.scdn.co/image/ab6761610000e5eb5d0fc4...   35379200          88   
6   https://i.scdn.co/image/ab6761610000e5eb41c3d8...    7420265          82   
7   https://i.scdn.co/image/b040846ceba13c3e9c125d...   42889908          84   
8   https://i.scdn.co/image/ab6761610000e5eb451eb0...    4270893          72   
9   https://i.scdn.co/image/84282c28d851a700132356...   16011536          81   
10  https://i.scdn.co/image/207803ce008388d3427a68...   12588906          77   
11  https://i.scdn.co/image/ab6761610000e5ebd8b998...   69271495          89   
12  https://i.scdn.co/image/ab6761610000e5ebadd503...   17967112          85   
13  https://i.scdn.co/image/ab6761610000e5eb019d68...   502

In [88]:
artist_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   artist_id     20 non-null     object
 1   artist_name   20 non-null     object
 2   external_url  20 non-null     object
 3   genre         20 non-null     object
 4   image_url     20 non-null     object
 5   followers     20 non-null     int64 
 6   popularity    20 non-null     int64 
 7   type          20 non-null     object
 8   artist_uri    20 non-null     object
dtypes: int64(2), object(7)
memory usage: 1.5+ KB


<h3>All of our data types are correct and no null values.</h3>


In [89]:
check_if_valid_data(artist_df)

True

<h1>ALBUMS</h1>

<h3>I would get all the artist ids from the pandas dataframe</h3>

In [90]:
artist_ids = artist_df['artist_id'].tolist()
artist_ids

['3TVXtAsR1Inumwj472S9r4',
 '4obzFoKoKRHIphyHzJ35G3',
 '2LIk90788K0zvyj2JJVwkJ',
 '790FomKkXshlbRYZFtlgla',
 '4q3ewBCX7sLwd24euuV69X',
 '1vyhD5VmyZ7KMfW5gqLgo5',
 '1dKdetem2xEmjgvyymzytS',
 '1dfeR4HaWDbWqFHLkxsg1d',
 '4MVyzYMgTwdP7Z49wAZHx0',
 '6olE6TJLqED3rqDCT0FyPh',
 '36QJpDe2go2KgaRleHCDTp',
 '6qqNVTkY8uBg9cP3Jd7DAH',
 '6l3HvQ5sa6mXTsMTB19rO5',
 '5pKCCKE2ajJHZ9KAiaK11H',
 '7jy3rLJdDQY21OgRLCZ9sD',
 '1GDbiv3spRmZ1XdM1jQbT7',
 '5M52tdBnJaKSvOpJGz8mfZ',
 '2cnMpRsOVqtPMfq7YiFE6K',
 '0k17h0D3J5VfsdmQ1iZtE9',
 '26dSoYclwsYLMAKD3tpOr4']

In [91]:
album_response = []

# Looping through all the artist_ids
for artist_id in artist_ids:

    # Pulling data via spotipy with an increased limit to get all the artists albums
    albums = sp.artist_albums(artist_id, album_type="album", limit=50, country="US")

    # Looping through all the albums available 
    for album in albums["items"]:
        # Inserting the data into a dictionary
        album_info = {
            "album_id": album["id"],
            "album_name": album["name"],
            "external_url": album["external_urls"]["spotify"],
            "image_url": album["images"][0]["url"],
            "release_date": album["release_date"],
            "total_tracks": album["total_tracks"],
            "type": album["type"],
            "album_uri": album["uri"],
            "artist_id": album["artists"][0]["id"],
        }
        album_response.append(album_info)


album_df = pd.DataFrame.from_dict(album_response)
album_df



album_id                                      album_name  \
0    3cf4iSSKd8ffTncbtKljXw                             Honestly, Nevermind   
1    2hicq6IKmDsPxZV7fJablX                             Honestly, Nevermind   
2    3SpBlxme9WbeQdI9kx7KAV                             Certified Lover Boy   
3    6sp02aeyiwfX35xRqwNiPv                             Certified Lover Boy   
4    6OQ9gBfg5EXeNAEwGSs6jK                            Dark Lane Demo Tapes   
..                      ...                                             ...   
325  1ePkYcH5ZQCb1b4tQeiEDj                                        Blackout   
326  0z7pVBGOD7HCIB7S8eLkLI                                     In The Zone   
327  5ax3GTsfX5uCUaNgnJsSG5                Britney (Digital Deluxe Version)   
328  5PmgtkodFl2Om3hMXONDll                         Oops!... I Did It Again   
329  3WNxdumkSMGMJRhEgK80qx  ...Baby One More Time (Digital Deluxe Version)   

                                          external_url  \
0    https://open.spotify.com/album/3cf4iSSKd8ffTnc...   
1    https://open.spotify.com/album/2hicq6IKmDsPxZV...   
2    https://open.spotify.com/album/3SpBlxme9WbeQdI...   
3    https://open.spotify.com/album/6sp02aeyiwfX35x...   
4    https://open.spotify.com/album/6OQ9gBfg5EXeNAE...   
..                                                 ...   
325  https://open.spotify.com/album/1ePkYcH5ZQCb1b4...   
326  https://open.spotify.com/album/0z7pVBGOD7HCIB7...   
327  https://open.spotify.com/album/5ax3GTsfX5uCUaN...   
328  https://open.spotify.com/album/5PmgtkodFl2Om3h...   
329  https://open.spotify.com/album/3WNxdumkSMGMJRh...   

                                             image_url release_date  \
0    https://i.scdn.co/image/ab67616d0000b2738dc0d8...   2022-06-17   
1    https://i.scdn.co/image/ab67616d0000b27322dfcf...   2022-06-16   
2    https://i.scdn.co/image/ab67616d0000b273cd945b...   2021-09-03   
3    https://i.scdn.co/image/ab67616d0000b2737529f7...   2021-09-02   
4    https://i.scdn.co/image/ab67616d0000b273bba7cf...   2020-05-01   
..                                                 ...          ...   
325  https://i.scdn.co/image/ab67616d0000b273ca10fa...   2007-10-25   
326  https://i.scdn.co/image/ab67616d0000b273efc698...   2003-11-13   
327  https://i.scdn.co/image/ab67616d0000b273e1a4e0...   2001-10-31   
328  https://i.scdn.co/image/ab67616d0000b2732aa206...   2000-05-16   
329  https://i.scdn.co/image/ab67616d0000b2738e4986...   1999-01-12   

     total_tracks   type                             album_uri  \
0              14  album  spotify:album:3cf4iSSKd8ffTncbtKljXw   
1              14  album  spotify:album:2hicq6IKmDsPxZV7fJablX   
2              21  album  spotify:album:3SpBlxme9WbeQdI9kx7KAV   
3              21  album  spotify:album:6sp02aeyiwfX35xRqwNiPv   
4              14  album  spotify:album:6OQ9gBfg5EXeNAEwGSs6jK   
..            ...    ...                                   ...   
325            15  album  spotify:album:1ePkYcH5ZQCb1b4tQeiEDj   
326            13  album  spotify:album:0z7pVBGOD7HCIB7S8eLkLI   
327            15  album  spotify:album:5ax3GTsfX5uCUaNgnJsSG5   
328            12  album  spotify:album:5PmgtkodFl2Om3hMXONDll   
329            16  album  spotify:album:3WNxdumkSMGMJRhEgK80qx   

                  artist_id  
0    3TVXtAsR1Inumwj472S9r4  
1    3TVXtAsR1Inumwj472S9r4  
2    3TVXtAsR1Inumwj472S9r4  
3    3TVXtAsR1Inumwj472S9r4  
4    3TVXtAsR1Inumwj472S9r4  
..                      ...  
325  26dSoYclwsYLMAKD3tpOr4  
326  26dSoYclwsYLMAKD3tpOr4  
327  26dSoYclwsYLMAKD3tpOr4  
328  26dSoYclwsYLMAKD3tpOr4  
329  26dSoYclwsYLMAKD3tpOr4  

[330 rows x 9 columns]

In [92]:
# remove duplicates 

album_df.drop_duplicates(subset=["album_name", "artist_id"], inplace=True)
album_df

album_id                                      album_name  \
0    3cf4iSSKd8ffTncbtKljXw                             Honestly, Nevermind   
2    3SpBlxme9WbeQdI9kx7KAV                             Certified Lover Boy   
4    6OQ9gBfg5EXeNAEwGSs6jK                            Dark Lane Demo Tapes   
6    7dqpveMVcWgbzqYrOdkFTD                                    Care Package   
8    2podUJIFG8hLfFz7Kqe8yJ                                     So Far Gone   
..                      ...                                             ...   
325  1ePkYcH5ZQCb1b4tQeiEDj                                        Blackout   
326  0z7pVBGOD7HCIB7S8eLkLI                                     In The Zone   
327  5ax3GTsfX5uCUaNgnJsSG5                Britney (Digital Deluxe Version)   
328  5PmgtkodFl2Om3hMXONDll                         Oops!... I Did It Again   
329  3WNxdumkSMGMJRhEgK80qx  ...Baby One More Time (Digital Deluxe Version)   

                                          external_url  \
0    https://open.spotify.com/album/3cf4iSSKd8ffTnc...   
2    https://open.spotify.com/album/3SpBlxme9WbeQdI...   
4    https://open.spotify.com/album/6OQ9gBfg5EXeNAE...   
6    https://open.spotify.com/album/7dqpveMVcWgbzqY...   
8    https://open.spotify.com/album/2podUJIFG8hLfFz...   
..                                                 ...   
325  https://open.spotify.com/album/1ePkYcH5ZQCb1b4...   
326  https://open.spotify.com/album/0z7pVBGOD7HCIB7...   
327  https://open.spotify.com/album/5ax3GTsfX5uCUaN...   
328  https://open.spotify.com/album/5PmgtkodFl2Om3h...   
329  https://open.spotify.com/album/3WNxdumkSMGMJRh...   

                                             image_url release_date  \
0    https://i.scdn.co/image/ab67616d0000b2738dc0d8...   2022-06-17   
2    https://i.scdn.co/image/ab67616d0000b273cd945b...   2021-09-03   
4    https://i.scdn.co/image/ab67616d0000b273bba7cf...   2020-05-01   
6    https://i.scdn.co/image/ab67616d0000b2739c1e02...   2019-08-02   
8    https://i.scdn.co/image/ab67616d0000b27397639c...   2019-02-14   
..                                                 ...          ...   
325  https://i.scdn.co/image/ab67616d0000b273ca10fa...   2007-10-25   
326  https://i.scdn.co/image/ab67616d0000b273efc698...   2003-11-13   
327  https://i.scdn.co/image/ab67616d0000b273e1a4e0...   2001-10-31   
328  https://i.scdn.co/image/ab67616d0000b2732aa206...   2000-05-16   
329  https://i.scdn.co/image/ab67616d0000b2738e4986...   1999-01-12   

     total_tracks   type                             album_uri  \
0              14  album  spotify:album:3cf4iSSKd8ffTncbtKljXw   
2              21  album  spotify:album:3SpBlxme9WbeQdI9kx7KAV   
4              14  album  spotify:album:6OQ9gBfg5EXeNAEwGSs6jK   
6              17  album  spotify:album:7dqpveMVcWgbzqYrOdkFTD   
8              18  album  spotify:album:2podUJIFG8hLfFz7Kqe8yJ   
..            ...    ...                                   ...   
325            15  album  spotify:album:1ePkYcH5ZQCb1b4tQeiEDj   
326            13  album  spotify:album:0z7pVBGOD7HCIB7S8eLkLI   
327            15  album  spotify:album:5ax3GTsfX5uCUaNgnJsSG5   
328            12  album  spotify:album:5PmgtkodFl2Om3hMXONDll   
329            16  album  spotify:album:3WNxdumkSMGMJRhEgK80qx   

                  artist_id  
0    3TVXtAsR1Inumwj472S9r4  
2    3TVXtAsR1Inumwj472S9r4  
4    3TVXtAsR1Inumwj472S9r4  
6    3TVXtAsR1Inumwj472S9r4  
8    3TVXtAsR1Inumwj472S9r4  
..                      ...  
325  26dSoYclwsYLMAKD3tpOr4  
326  26dSoYclwsYLMAKD3tpOr4  
327  26dSoYclwsYLMAKD3tpOr4  
328  26dSoYclwsYLMAKD3tpOr4  
329  26dSoYclwsYLMAKD3tpOr4  

[287 rows x 9 columns]

<h3>It removed 43 duplicate albums. Now I am going to remove deluxe albums.</h3>

In [93]:
album_df = album_df[album_df["album_name"].str.contains("(Deluxe|Deluxe)|Deluxe Edition|Deluxe Version|Super Deluxe") == False]

album_df

/var/folders/z_/d0nn0q354fb06flql_0rbx_80000gn/T/ipykernel_1789/3297947759.py:1: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  album_df = album_df[album_df["album_name"].str.contains("(Deluxe|Deluxe)|Deluxe Edition|Deluxe Version|Super Deluxe") == False]


album_id                      album_name  \
0    3cf4iSSKd8ffTncbtKljXw             Honestly, Nevermind   
2    3SpBlxme9WbeQdI9kx7KAV             Certified Lover Boy   
4    6OQ9gBfg5EXeNAEwGSs6jK            Dark Lane Demo Tapes   
6    7dqpveMVcWgbzqYrOdkFTD                    Care Package   
8    2podUJIFG8hLfFz7Kqe8yJ                     So Far Gone   
..                      ...                             ...   
318  2vnJKtGjZXRUg0mYPZ3HGH          A Saucerful of Secrets   
319  2Se4ZylF9NkFGD92yv1aZC  The Piper at the Gates of Dawn   
325  1ePkYcH5ZQCb1b4tQeiEDj                        Blackout   
326  0z7pVBGOD7HCIB7S8eLkLI                     In The Zone   
328  5PmgtkodFl2Om3hMXONDll         Oops!... I Did It Again   

                                          external_url  \
0    https://open.spotify.com/album/3cf4iSSKd8ffTnc...   
2    https://open.spotify.com/album/3SpBlxme9WbeQdI...   
4    https://open.spotify.com/album/6OQ9gBfg5EXeNAE...   
6    https://open.spotify.com/album/7dqpveMVcWgbzqY...   
8    https://open.spotify.com/album/2podUJIFG8hLfFz...   
..                                                 ...   
318  https://open.spotify.com/album/2vnJKtGjZXRUg0m...   
319  https://open.spotify.com/album/2Se4ZylF9NkFGD9...   
325  https://open.spotify.com/album/1ePkYcH5ZQCb1b4...   
326  https://open.spotify.com/album/0z7pVBGOD7HCIB7...   
328  https://open.spotify.com/album/5PmgtkodFl2Om3h...   

                                             image_url release_date  \
0    https://i.scdn.co/image/ab67616d0000b2738dc0d8...   2022-06-17   
2    https://i.scdn.co/image/ab67616d0000b273cd945b...   2021-09-03   
4    https://i.scdn.co/image/ab67616d0000b273bba7cf...   2020-05-01   
6    https://i.scdn.co/image/ab67616d0000b2739c1e02...   2019-08-02   
8    https://i.scdn.co/image/ab67616d0000b27397639c...   2019-02-14   
..                                                 ...          ...   
318  https://i.scdn.co/image/ab67616d0000b2734b4fcf...   1968-06-29   
319  https://i.scdn.co/image/ab67616d0000b273474c75...   1967-08-05   
325  https://i.scdn.co/image/ab67616d0000b273ca10fa...   2007-10-25   
326  https://i.scdn.co/image/ab67616d0000b273efc698...   2003-11-13   
328  https://i.scdn.co/image/ab67616d0000b2732aa206...   2000-05-16   

     total_tracks   type                             album_uri  \
0              14  album  spotify:album:3cf4iSSKd8ffTncbtKljXw   
2              21  album  spotify:album:3SpBlxme9WbeQdI9kx7KAV   
4              14  album  spotify:album:6OQ9gBfg5EXeNAEwGSs6jK   
6              17  album  spotify:album:7dqpveMVcWgbzqYrOdkFTD   
8              18  album  spotify:album:2podUJIFG8hLfFz7Kqe8yJ   
..            ...    ...                                   ...   
318             7  album  spotify:album:2vnJKtGjZXRUg0mYPZ3HGH   
319            11  album  spotify:album:2Se4ZylF9NkFGD92yv1aZC   
325            15  album  spotify:album:1ePkYcH5ZQCb1b4tQeiEDj   
326            13  album  spotify:album:0z7pVBGOD7HCIB7S8eLkLI   
328            12  album  spotify:album:5PmgtkodFl2Om3hMXONDll   

                  artist_id  
0    3TVXtAsR1Inumwj472S9r4  
2    3TVXtAsR1Inumwj472S9r4  
4    3TVXtAsR1Inumwj472S9r4  
6    3TVXtAsR1Inumwj472S9r4  
8    3TVXtAsR1Inumwj472S9r4  
..                      ...  
318  0k17h0D3J5VfsdmQ1iZtE9  
319  0k17h0D3J5VfsdmQ1iZtE9  
325  26dSoYclwsYLMAKD3tpOr4  
326  26dSoYclwsYLMAKD3tpOr4  
328  26dSoYclwsYLMAKD3tpOr4  

[237 rows x 9 columns]

<h3>237 albums total. 50 deluxe albums deleted. A total of 93 albums were deleted. </h3>

In [94]:
album_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 237 entries, 0 to 328
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   album_id      237 non-null    object
 1   album_name    237 non-null    object
 2   external_url  237 non-null    object
 3   image_url     237 non-null    object
 4   release_date  237 non-null    object
 5   total_tracks  237 non-null    int64 
 6   type          237 non-null    object
 7   album_uri     237 non-null    object
 8   artist_id     237 non-null    object
dtypes: int64(1), object(8)
memory usage: 18.5+ KB


<h3>No null values, we have all our data types correct except for release_date</h3>

In [95]:
album_df['release_date'] = pd.to_datetime(album_df["release_date"], format="%Y-%m-%d")

/var/folders/z_/d0nn0q354fb06flql_0rbx_80000gn/T/ipykernel_1789/2820813450.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  album_df['release_date'] = pd.to_datetime(album_df["release_date"], format="%Y-%m-%d")


In [96]:
album_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 237 entries, 0 to 328
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   album_id      237 non-null    object        
 1   album_name    237 non-null    object        
 2   external_url  237 non-null    object        
 3   image_url     237 non-null    object        
 4   release_date  237 non-null    datetime64[ns]
 5   total_tracks  237 non-null    int64         
 6   type          237 non-null    object        
 7   album_uri     237 non-null    object        
 8   artist_id     237 non-null    object        
dtypes: datetime64[ns](1), int64(1), object(7)
memory usage: 18.5+ KB


<h3>Perfect!</h3>

In [97]:
check_if_valid_data(album_df)

True

<h1>TRACKS</h1>

In [98]:
album_ids = album_df['album_id'].tolist()
album_ids

['3cf4iSSKd8ffTncbtKljXw',
 '3SpBlxme9WbeQdI9kx7KAV',
 '6OQ9gBfg5EXeNAEwGSs6jK',
 '7dqpveMVcWgbzqYrOdkFTD',
 '2podUJIFG8hLfFz7Kqe8yJ',
 '1ATL5GLyefJaxhQzSPVrLX',
 '1lXY618HWkwYKJWBRYR4MK',
 '40GMAhriYJRO1rsY4YdrZb',
 '15QCBYjP6HwHvsff100UBx',
 '0ptlfJfwGTy0Yvrk14JK1I',
 '1XslIirSxfAhhxRdn4Li9t',
 '6jlrjFR9mJV3jd1IPSplXU',
 '7eC4wtMG1I2Jtk4FDWbkKC',
 '13MHW8hoLFjX7SaVEVmj3X',
 '2eE8BVirX9VF8Di9hD90iw',
 '5g2INX6dnMIJG0xKygXwF3',
 '7AaqMMiYMvnMB3RcS8u3EY',
 '7xfcihbsBamUi6sezd0oOG',
 '50UV94JhFbWhF37n74Jhk3',
 '0txbd4bmIM8Hq94jBrS0lc',
 '2Qvup4Tfqt5r7icH44mx9s',
 '5CS8E6JVACItYto4OOJoPW',
 '4i5b4YWuMtneUSvQPONwzK',
 '12nlJpvrOd7tTOaCxB1UeR',
 '3RQQmkQEvNCY4prGKE6oc5',
 '2d9BCZeAAhiZWPpbX9aPCW',
 '4gvQO5mEuhbMSrLIuwXkmz',
 '5lJqux7orBlA1QzyiBGti1',
 '6ylFfzx32ICw4L1A7YWNLN',
 '7CjJb2mikwAWA1V6kewFBF',
 '11GmvpYnbgK0rSryPaV5BP',
 '2mX8ktJoWvyidWBU9U8Jis',
 '5kprdYds6oZb4iSldfflOT',
 '79447qgPdAlZVX0Xx4Fzfv',
 '4cGc9Eeb3Gjff2Aq5ILLEf',
 '4loQMkHOAbqVdyb59KBbsp',
 '5Fqdd85mAoGDp0K9lmbPDp',
 

In [99]:
track_response = []

# Going through each album id and searching for it's tracks
for album_id in album_ids:
    tracks = sp.album_tracks(album_id)

    # Adding data to dictionary 
    for track in tracks["items"]:
        track_info = {
            "track_id": track["id"],
            "song_name": track["name"],
            "external_url": track["external_urls"]["spotify"],
            "duration_ms": track["duration_ms"],
            "explicit": track["explicit"],
            "disc_number": track["disc_number"],
            "type": track["type"],
            "song_uri": track["uri"],
            "album_id": album_id,
        }

        track_response.append(track_info)

track_df = pd.DataFrame.from_dict(track_response)
track_df

track_id               song_name  \
0     1kNdtonJXHorsVpWerK8C2                   Intro   
1     1vbn9fEyw1IYhqgZJdu9ZB            Falling Back   
2     4FIgHqXd4KkOwxrGeLDYM2          Texts Go Green   
3     1eCFz60zd7mAXgWLapPd9B                Currents   
4     0nAZGkBGKQCXyaoSJfRhC1                A Keeper   
...                      ...                     ...   
3379  5iW9Y4seDbBgIchT0ANY3q       One Kiss from You   
3380  4MOBPIIRUgHTBAp0NpgLTm       Where Are You Now   
3381  7pY6xLJqmezJNG7VwuKVSH  Can't Make You Love Me   
3382  30hSG0Is8yOGkFQtF20Lxu   When Your Eyes Say It   
3383  1NmjzfxyAN9u5EaqHk1uIY              Dear Diary   

                                           external_url  duration_ms  \
0     https://open.spotify.com/track/1kNdtonJXHorsVp...        36935   
1     https://open.spotify.com/track/1vbn9fEyw1IYhqg...       266178   
2     https://open.spotify.com/track/4FIgHqXd4KkOwxr...       308820   
3     https://open.spotify.com/track/1eCFz60zd7mAXgW...       157390   
4     https://open.spotify.com/track/0nAZGkBGKQCXyao...       173097   
...                                                 ...          ...   
3379  https://open.spotify.com/track/5iW9Y4seDbBgIch...       203280   
3380  https://open.spotify.com/track/4MOBPIIRUgHTBAp...       279386   
3381  https://open.spotify.com/track/7pY6xLJqmezJNG7...       197186   
3382  https://open.spotify.com/track/30hSG0Is8yOGkFQ...       269773   
3383  https://open.spotify.com/track/1NmjzfxyAN9u5Ea...       166653   

      explicit  disc_number   type                              song_uri  \
0        False            1  track  spotify:track:1kNdtonJXHorsVpWerK8C2   
1        False            1  track  spotify:track:1vbn9fEyw1IYhqgZJdu9ZB   
2        False            1  track  spotify:track:4FIgHqXd4KkOwxrGeLDYM2   
3        False            1  track  spotify:track:1eCFz60zd7mAXgWLapPd9B   
4         True            1  track  spotify:track:0nAZGkBGKQCXyaoSJfRhC1   
...        ...          ...    ...                                   ...   
3379     False            1  track  spotify:track:5iW9Y4seDbBgIchT0ANY3q   
3380     False            1  track  spotify:track:4MOBPIIRUgHTBAp0NpgLTm   
3381     False            1  track  spotify:track:7pY6xLJqmezJNG7VwuKVSH   
3382     False            1  track  spotify:track:30hSG0Is8yOGkFQtF20Lxu   
3383     False            1  track  spotify:track:1NmjzfxyAN9u5EaqHk1uIY   

                    album_id  
0     3cf4iSSKd8ffTncbtKljXw  
1     3cf4iSSKd8ffTncbtKljXw  
2     3cf4iSSKd8ffTncbtKljXw  
3     3cf4iSSKd8ffTncbtKljXw  
4     3cf4iSSKd8ffTncbtKljXw  
...                      ...  
3379  5PmgtkodFl2Om3hMXONDll  
3380  5PmgtkodFl2Om3hMXONDll  
3381  5PmgtkodFl2Om3hMXONDll  
3382  5PmgtkodFl2Om3hMXONDll  
3383  5PmgtkodFl2Om3hMXONDll  

[3384 rows x 9 columns]

In [100]:
track_df.drop_duplicates(subset=["song_name", "album_id"], inplace=True)
track_df

track_id               song_name  \
0     1kNdtonJXHorsVpWerK8C2                   Intro   
1     1vbn9fEyw1IYhqgZJdu9ZB            Falling Back   
2     4FIgHqXd4KkOwxrGeLDYM2          Texts Go Green   
3     1eCFz60zd7mAXgWLapPd9B                Currents   
4     0nAZGkBGKQCXyaoSJfRhC1                A Keeper   
...                      ...                     ...   
3379  5iW9Y4seDbBgIchT0ANY3q       One Kiss from You   
3380  4MOBPIIRUgHTBAp0NpgLTm       Where Are You Now   
3381  7pY6xLJqmezJNG7VwuKVSH  Can't Make You Love Me   
3382  30hSG0Is8yOGkFQtF20Lxu   When Your Eyes Say It   
3383  1NmjzfxyAN9u5EaqHk1uIY              Dear Diary   

                                           external_url  duration_ms  \
0     https://open.spotify.com/track/1kNdtonJXHorsVp...        36935   
1     https://open.spotify.com/track/1vbn9fEyw1IYhqg...       266178   
2     https://open.spotify.com/track/4FIgHqXd4KkOwxr...       308820   
3     https://open.spotify.com/track/1eCFz60zd7mAXgW...       157390   
4     https://open.spotify.com/track/0nAZGkBGKQCXyao...       173097   
...                                                 ...          ...   
3379  https://open.spotify.com/track/5iW9Y4seDbBgIch...       203280   
3380  https://open.spotify.com/track/4MOBPIIRUgHTBAp...       279386   
3381  https://open.spotify.com/track/7pY6xLJqmezJNG7...       197186   
3382  https://open.spotify.com/track/30hSG0Is8yOGkFQ...       269773   
3383  https://open.spotify.com/track/1NmjzfxyAN9u5Ea...       166653   

      explicit  disc_number   type                              song_uri  \
0        False            1  track  spotify:track:1kNdtonJXHorsVpWerK8C2   
1        False            1  track  spotify:track:1vbn9fEyw1IYhqgZJdu9ZB   
2        False            1  track  spotify:track:4FIgHqXd4KkOwxrGeLDYM2   
3        False            1  track  spotify:track:1eCFz60zd7mAXgWLapPd9B   
4         True            1  track  spotify:track:0nAZGkBGKQCXyaoSJfRhC1   
...        ...          ...    ...                                   ...   
3379     False            1  track  spotify:track:5iW9Y4seDbBgIchT0ANY3q   
3380     False            1  track  spotify:track:4MOBPIIRUgHTBAp0NpgLTm   
3381     False            1  track  spotify:track:7pY6xLJqmezJNG7VwuKVSH   
3382     False            1  track  spotify:track:30hSG0Is8yOGkFQtF20Lxu   
3383     False            1  track  spotify:track:1NmjzfxyAN9u5EaqHk1uIY   

                    album_id  
0     3cf4iSSKd8ffTncbtKljXw  
1     3cf4iSSKd8ffTncbtKljXw  
2     3cf4iSSKd8ffTncbtKljXw  
3     3cf4iSSKd8ffTncbtKljXw  
4     3cf4iSSKd8ffTncbtKljXw  
...                      ...  
3379  5PmgtkodFl2Om3hMXONDll  
3380  5PmgtkodFl2Om3hMXONDll  
3381  5PmgtkodFl2Om3hMXONDll  
3382  5PmgtkodFl2Om3hMXONDll  
3383  5PmgtkodFl2Om3hMXONDll  

[3382 rows x 9 columns]

<h3>It removed 2 songs</h3>

In [101]:
track_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3382 entries, 0 to 3383
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   track_id      3382 non-null   object
 1   song_name     3382 non-null   object
 2   external_url  3382 non-null   object
 3   duration_ms   3382 non-null   int64 
 4   explicit      3382 non-null   bool  
 5   disc_number   3382 non-null   int64 
 6   type          3382 non-null   object
 7   song_uri      3382 non-null   object
 8   album_id      3382 non-null   object
dtypes: bool(1), int64(2), object(6)
memory usage: 241.1+ KB


<h3>All of our data types are correct and no null values.</h3>

In [102]:
check_if_valid_data(track_df)

True

<h1>TRACK FEATURES</h1>

In [103]:
track_ids = track_df["track_id"].tolist()
track_ids

['1kNdtonJXHorsVpWerK8C2',
 '1vbn9fEyw1IYhqgZJdu9ZB',
 '4FIgHqXd4KkOwxrGeLDYM2',
 '1eCFz60zd7mAXgWLapPd9B',
 '0nAZGkBGKQCXyaoSJfRhC1',
 '7sT7kZEYd1MrmzLLIRVZas',
 '4rmVZajAF7PkrCagGPHbqa',
 '28JBD8p18xNuOfyV7Cotdn',
 '6Yj7Zhxt73uvwFFvzQXdxO',
 '7hRlTK3mgFTBEjX7DjqBsI',
 '3N3d4OWbPVxTQVBFqNyykk',
 '1XsRwo2it6QHdV2OiT6IzF',
 '2vy9Ry1TNjCD85TiCbTz0r',
 '3F5CgOj3wFlRv51JsHbxhe',
 '2HSmyk2qMN8WQjuGhaQgCk',
 '6jy9yJfgCsMHdu2Oz4BGKX',
 '37Nqx7iavZpotJSDXZWbJ3',
 '61S79KIVA4I9FXbnsylEHT',
 '4VCbgIdr8ptegWeJpqLVHH',
 '40iJIUlhi6renaREYGeIDS',
 '0k1WUmIRnG3xU6fvvDVfRG',
 '4s7QLoImIwmPi9L6dq1nVW',
 '34D6FJysnQioVingDKufuf',
 '11pEKMLmavDu8fxOB5QjbQ',
 '1PDP7mLiAMwhfmgIwzhOm2',
 '2tUL6dZf1mywCj5WvCPZw6',
 '2BcMwX1MPV6ZHP4tUT9uq6',
 '42m3eP1JJhtzffal9B136J',
 '2Q3jFbyE61mCjS3SkW4toJ',
 '08XWh5c0BMyD1nKVxxl91z',
 '5ScbulRnixQ2XAdvrPMFjz',
 '50ceCGZ3oD3U5caQV5bP6f',
 '7F9cT6hIRhnFCYP6GKS0tf',
 '3RkNXZvOSMMElmmXztDc94',
 '1cCXT1cErAY6tmIqHbodrt',
 '3IvMYBE7A3c7to1aEcfFJk',
 '5TCBWmEBrin7etRa4Lswr1',
 

In [104]:
num_of_tracks = len(track_ids)

starting_idx = 0
stopping_idx = 100

feature_response = []

# Loop to make API calls with a maximum of 100 tracks at a time 
for i in range((num_of_tracks // 100) + 1):
    features = sp.audio_features(track_ids[starting_idx:stopping_idx])

    # Add data to the dictionary 
    for feature in features:
        feature_info = {
            "track_id": feature["id"],
            "danceability": feature["danceability"],
            "energy": feature["energy"],
            "instrumentalness": feature["instrumentalness"],
            "liveness": feature["liveness"],
            "loudness": feature["loudness"],
            "speechiness": feature["speechiness"],
            "tempo": feature["tempo"],
            "type": feature["type"],
            "valence": feature["valence"],
            "song_uri": feature["uri"],
        }

        feature_response.append(feature_info)
    
    # Increase our indexes and move the window up 
    starting_idx = stopping_idx
    stopping_idx += 100

feature_df = pd.DataFrame.from_dict(feature_response)
feature_df

track_id  danceability  energy  instrumentalness  \
0     1kNdtonJXHorsVpWerK8C2         0.193  0.0218          0.880000   
1     1vbn9fEyw1IYhqgZJdu9ZB         0.718  0.7580          0.006530   
2     4FIgHqXd4KkOwxrGeLDYM2         0.765  0.6270          0.195000   
3     1eCFz60zd7mAXgWLapPd9B         0.883  0.2830          0.001670   
4     0nAZGkBGKQCXyaoSJfRhC1         0.600  0.4820          0.000104   
...                      ...           ...     ...               ...   
3377  5iW9Y4seDbBgIchT0ANY3q         0.732  0.8880          0.000003   
3378  4MOBPIIRUgHTBAp0NpgLTm         0.636  0.4890          0.000000   
3379  7pY6xLJqmezJNG7VwuKVSH         0.810  0.9240          0.000087   
3380  30hSG0Is8yOGkFQtF20Lxu         0.693  0.6250          0.000000   
3381  1NmjzfxyAN9u5EaqHk1uIY         0.509  0.1570          0.000000   

      liveness  loudness  speechiness    tempo            type  valence  \
0       0.1170   -31.160       0.0389  125.501  audio_features    0.213   
1       0.1100    -8.290       0.0703  118.989  audio_features    0.349   
2       0.0847    -4.607       0.0622  118.997  audio_features    0.536   
3       0.0935   -12.653       0.1130  122.013  audio_features    0.416   
4       0.1150   -11.596       0.0701  121.909  audio_features    0.323   
...        ...       ...          ...      ...             ...      ...   
3377    0.3790    -2.415       0.0792   93.991  audio_features    0.674   
3378    0.2910    -8.098       0.0263  125.616  audio_features    0.297   
3379    0.1490    -4.020       0.0356  108.016  audio_features    0.868   
3380    0.6590    -7.147       0.0431   86.040  audio_features    0.439   
3381    0.1660   -14.285       0.0333  127.660  audio_features    0.288   

                                  song_uri  
0     spotify:track:1kNdtonJXHorsVpWerK8C2  
1     spotify:track:1vbn9fEyw1IYhqgZJdu9ZB  
2     spotify:track:4FIgHqXd4KkOwxrGeLDYM2  
3     spotify:track:1eCFz60zd7mAXgWLapPd9B  
4     spotify:track:0nAZGkBGKQCXyaoSJfRhC1  
...                                    ...  
3377  spotify:track:5iW9Y4seDbBgIchT0ANY3q  
3378  spotify:track:4MOBPIIRUgHTBAp0NpgLTm  
3379  spotify:track:7pY6xLJqmezJNG7VwuKVSH  
3380  spotify:track:30hSG0Is8yOGkFQtF20Lxu  
3381  spotify:track:1NmjzfxyAN9u5EaqHk1uIY  

[3382 rows x 11 columns]

In [105]:
feature_df.drop_duplicates(subset="track_id", inplace=True)
feature_df


track_id  danceability  energy  instrumentalness  \
0     1kNdtonJXHorsVpWerK8C2         0.193  0.0218          0.880000   
1     1vbn9fEyw1IYhqgZJdu9ZB         0.718  0.7580          0.006530   
2     4FIgHqXd4KkOwxrGeLDYM2         0.765  0.6270          0.195000   
3     1eCFz60zd7mAXgWLapPd9B         0.883  0.2830          0.001670   
4     0nAZGkBGKQCXyaoSJfRhC1         0.600  0.4820          0.000104   
...                      ...           ...     ...               ...   
3377  5iW9Y4seDbBgIchT0ANY3q         0.732  0.8880          0.000003   
3378  4MOBPIIRUgHTBAp0NpgLTm         0.636  0.4890          0.000000   
3379  7pY6xLJqmezJNG7VwuKVSH         0.810  0.9240          0.000087   
3380  30hSG0Is8yOGkFQtF20Lxu         0.693  0.6250          0.000000   
3381  1NmjzfxyAN9u5EaqHk1uIY         0.509  0.1570          0.000000   

      liveness  loudness  speechiness    tempo            type  valence  \
0       0.1170   -31.160       0.0389  125.501  audio_features    0.213   
1       0.1100    -8.290       0.0703  118.989  audio_features    0.349   
2       0.0847    -4.607       0.0622  118.997  audio_features    0.536   
3       0.0935   -12.653       0.1130  122.013  audio_features    0.416   
4       0.1150   -11.596       0.0701  121.909  audio_features    0.323   
...        ...       ...          ...      ...             ...      ...   
3377    0.3790    -2.415       0.0792   93.991  audio_features    0.674   
3378    0.2910    -8.098       0.0263  125.616  audio_features    0.297   
3379    0.1490    -4.020       0.0356  108.016  audio_features    0.868   
3380    0.6590    -7.147       0.0431   86.040  audio_features    0.439   
3381    0.1660   -14.285       0.0333  127.660  audio_features    0.288   

                                  song_uri  
0     spotify:track:1kNdtonJXHorsVpWerK8C2  
1     spotify:track:1vbn9fEyw1IYhqgZJdu9ZB  
2     spotify:track:4FIgHqXd4KkOwxrGeLDYM2  
3     spotify:track:1eCFz60zd7mAXgWLapPd9B  
4     spotify:track:0nAZGkBGKQCXyaoSJfRhC1  
...                                    ...  
3377  spotify:track:5iW9Y4seDbBgIchT0ANY3q  
3378  spotify:track:4MOBPIIRUgHTBAp0NpgLTm  
3379  spotify:track:7pY6xLJqmezJNG7VwuKVSH  
3380  spotify:track:30hSG0Is8yOGkFQtF20Lxu  
3381  spotify:track:1NmjzfxyAN9u5EaqHk1uIY  

[3382 rows x 11 columns]

<h3>No duplicates and there are the same amount of tracks as features.</h3>

In [106]:
feature_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3382 entries, 0 to 3381
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   track_id          3382 non-null   object 
 1   danceability      3382 non-null   float64
 2   energy            3382 non-null   float64
 3   instrumentalness  3382 non-null   float64
 4   liveness          3382 non-null   float64
 5   loudness          3382 non-null   float64
 6   speechiness       3382 non-null   float64
 7   tempo             3382 non-null   float64
 8   type              3382 non-null   object 
 9   valence           3382 non-null   float64
 10  song_uri          3382 non-null   object 
dtypes: float64(8), object(3)
memory usage: 317.1+ KB


<h3>All of our data types are correct and no null values.</h3>

In [107]:
check_if_valid_data(feature_df)

True

<h1>LOAD INTO DATABASE</h1>

In [108]:

# Create a database connection
try:
    conn = sqlite3.connect("../spotify.db")
except Error as error:
    print(error)


In [109]:
artist_df.to_sql('artist', conn, index=False, if_exists="append")
album_df.to_sql('album', conn, index=False, if_exists="append")
track_df.to_sql('track', conn, index=False, if_exists="append")
feature_df.to_sql('track_feature', conn, index=False, if_exists="append")

3382

In [110]:
conn.close()